vault/src/execution.rs

```diff
@@ -237,16 +239,135 @@ impl Request {
+
+                    let ret: Result<u64, _> = x
+                        .into_inner()
+                        .checked_div(FixedU128::accuracy())
+                        .ok_or(Error::ArithmeticUnderflow)
+                        .and_then(|x| x.try_into().map_err(Into::<Error>::into));
+                    futures::future::ready(ret)
+                })
+                .try_filter_map(|x| async move {
+                    match btc_rpc.fee_rate(txid) {
+                        Ok(current_fee) => {
+                            if x >= current_fee {
+                                Ok(Some((current_fee, x)))
+                            } else {
+                                Ok(None)
+                            }
+                        }
+                        Err(e) => {
+                            tracing::debug!("Failed to get fee_rate: {}", e);
+                            Ok(None)
+                        }
+                    }
+                })
+                .filter_map(|x| async {
+                    match btc_rpc.is_in_mempool(txid_copy) {
+                        Ok(false) => {
+                            //   if not in mempool anymore, don't propagate the event (even if it is an error)
+                            tracing::debug!("Txid not in mempool anymore...");
+                            None
+                        }
+                        Ok(true) => {
+                            tracing::debug!("Txid is still in mempool...");
+                            Some(x)
+                        }
+                        Err(e) => Some(Err(e.into())),
+                    }
+                });
+
+            let wait_for_transaction_metadata = btc_rpc.wait_for_transaction_metadata(txid, num_confirmations);
+            futures::pin_mut!(subscription);
+
+            let mut metadata_fut = wait_for_transaction_metadata;
+
+            // The code below looks a little bit complicated but the idea is simple:
+            // we keep waiting for inclusion until it's either included in the bitcoin chain,
+            // or we successfully bump fees
+            let tx_metadata = loop {
+                match futures::future::select(metadata_fut, subscription.next()).await {
+                    Either::Left((result, _)) => break result?,
+                    Either::Right((None, _)) => return Err(Error::RuntimeError(RuntimeError::ChannelClosed)),
+                    Either::Right((Some(Err(x)), continuation)) => {
+                        tracing::warn!("Received an error from the fee rate subscription: {x}");
+                        // continue with the unchanged fee rate
+                        metadata_fut = continuation;
+                    }
+                    Either::Right((Some(Ok((old_fee, new_fee))), continuation)) => {
+                        tracing::debug!("Attempting to bump fee rate from {old_fee} to {new_fee}...");
+                        match self.bump_fee(btc_rpc, txid, new_fee).await {
```

Comment:
Is it possible to end up in a sort of "deadlock" if the Vault is always increasing fees, say if we do receive numerous smaller fee estimate bumps on the parachain? (assuming re-submission pushes the new tx to the "back" of the mempool)

Reply:
I don't see how that'd be possible, no. If `bump_fee` succeeds, the new tx will end up more to the front of the queue because the miners sort by fee. If it fails there'd be no change